lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -699,6 +700,93 @@ FIRRTLType MemOp::getDataTypeOrNull() {
 // Statements
 //===----------------------------------------------------------------------===//
 
+static bool areConnectEquivalent(FIRRTLType destType, FIRRTLType srcType);
+
+/// Helper to implement the equivalence logic for a pair of bundle elements in a
+/// connect.
+static bool areElementsConnectEquivalent(
+    std::tuple<BundleType::BundleElement, BundleType::BundleElement> tuple) {
+  BundleType::BundleElement destElement = std::get<0>(tuple);
+  Identifier destElementName = std::get<0>(destElement);
+  FIRRTLType destElementType = std::get<1>(destElement);
+  bool destIsFlip = destElementType.isa<FlipType>();
+
+  BundleType::BundleElement srcElement = std::get<1>(tuple);
+  Identifier srcElementName = std::get<0>(srcElement);
+  FIRRTLType srcElementType = std::get<1>(srcElement);
+  bool srcIsFlip = srcElementType.isa<FlipType>();
+
+  if (destElementName != srcElementName)
+    return false;
+
+  if (destIsFlip != srcIsFlip)
+    return false;
+
+  // If either destination or source is flipped, strip the outer flip for
+  // recursively equivalance checking.
+  if (destIsFlip)
+    destElementType = destElementType.cast<FlipType>().getElementType();
+
+  if (srcIsFlip)
+    srcElementType = srcElementType.cast<FlipType>().getElementType();
+
+  return areConnectEquivalent(destElementType, srcElementType);
+}
+
+/// Returns whether the two types are equivalent for the purpose of
+/// connections. See Section 4.5 in the FIRRTL spec:
+/// https://github.com/freechipsproject/firrtl/blob/master/spec/spec.pdf.
+static bool areConnectEquivalent(FIRRTLType destType, FIRRTLType srcType) {
```

Comment:
Type equivalence is a generally defined FIRRTL thing, so I'd recommend splitting this out into a more general "areTypesEquivalent" predicate in FIRRTL/Types.cpp.

Reply:
I'd start this function out with:

destType = destType.getPassiveType();
srcType = srcType.getPassiveType();

because flips can occur at many levels and (AFAIK) don't affect equivalence. If you want to be fancy, do that once in the non-recursive public entrypoint to the predicate.

Reply:
Yep, this is going to be updated as you say. See my discussion in the [issue](https://github.com/llvm/circt/issues/232#issuecomment-733115018) for the reasoning that led me to the same conclusion.